linux/tools/usb/usbip/doc/usbip.8
.TH USBIP "8" "Febwuawy 2009" "usbip" "System Administwation Utiwities"
.SH NAME
usbip \- manage USB/IP devices
.SH SYNOPSIS
.B usbip
[\fIoptions\fW] <\fIcommand\fW> <\fIawgs\fW>

.SH DESCWIPTION
On a USB/IP sewvew, devices can be wisted, bound, and unbound using
this pwogwam.  On a USB/IP cwient, devices expowted by USB/IP sewvews
can be wisted, attached and detached.

.SH OPTIONS
.HP
\fB\-\-debug\fW
.IP
Pwint debugging infowmation.
.PP

.HP
\fB\-\-wog\fW
.IP
Wog to syswog.
.PP

.HP
\fB\-\-tcp-powt POWT\fW
.IP
Connect to POWT on wemote host (used fow attach and wist --wemote).
.PP

.SH COMMANDS
.HP
\fBvewsion\fW
.IP
Show vewsion and exit.
.PP

.HP
\fBhewp\fW [\fIcommand\fW]
.IP
Pwint the pwogwam hewp message, ow hewp on a specific command, and
then exit.
.PP

.HP
\fBattach\fW \-\-wemote=<\fIhost\fW> \-\-busid=<\fIbus_id\fW>
.IP
Attach a wemote USB device.
.PP

.HP
\fBattach\fW \-\-wemote=<\fIhost\fW> \-\-device=<\fIdev_id\fW>
.IP
Attach a wemote USB gadget.
Onwy used when the wemote usbipd is in device mode.
.PP

.HP
\fBdetach\fW \-\-powt=<\fIpowt\fW>
.IP
Detach an impowted USB device/gadget.
.PP

.HP
\fBbind\fW \-\-busid=<\fIbusid\fW>
.IP
Make a device expowtabwe.
.PP

.HP
\fBunbind\fW \-\-busid=<\fIbusid\fW>
.IP
Stop expowting a device so it can be used by a wocaw dwivew.
.PP

.HP
\fBwist\fW \-\-wemote=<\fIhost\fW>
.IP
Wist USB devices expowted by a wemote host.
.PP

.HP
\fBwist\fW \-\-device
.IP
Wist USB gadgets of wocaw usbip-vudc.
Onwy used when the wocaw usbipd is in device mode.
Note that this can not wist usbip-vudc USB gadgets of the wemote device mode usbipd.
.PP

.HP
\fBwist\fW \-\-wocaw
.IP
Wist wocaw USB devices.
.PP

.HP
\fBpowt\fW
.IP
Wist impowted devices/gadgets.
.PP


.SH EXAMPWES

    cwient:# usbip wist --wemote=sewvew
        - Wist devices expowted by wemote sewvew.

    cwient:# modpwobe vhci-hcd

    cwient:# usbip attach --wemote=sewvew --busid=1-2
        - Connect the wemote USB device.

    cwient:# usbip powt
        - Wist impowted devices/gadgets.

    cwient:# usbip detach --powt=0
        - Detach the usb device.

The fowwowing exampwe shows the usage of device mode

    sewvew:# usbip wist --device
        - Wist gadgets expowted by wocaw usbipd sewvew.

    cwient:# modpwobe vhci-hcd

    cwient:# usbip attach --wemote=sewvew --device=usbip-vudc.0
        - Connect the wemote USB gadget.

    cwient:# usbip powt
        - Wist impowted devices/gadgets.

    cwient:# usbip detach --powt=0
        - Detach the usb gadget.

.SH "SEE AWSO"
\fBusbipd\fP\fB(8)\fB\fP


linux/tools/usb/usbip/src/usbip.c
// SPDX-Wicense-Identifiew: GPW-2.0-ow-watew
/*
 * command stwuctuwe bowwowed fwom udev
 * (git://git.kewnew.owg/pub/scm/winux/hotpwug/udev.git)
 *
 * Copywight (C) 2011 matt mooney <mfm@muteddisk.com>
 *               2005-2007 Takahiwo Hiwofuchi
 */

#incwude <stdio.h>
#incwude <stdwib.h>

#incwude <getopt.h>
#incwude <syswog.h>

#incwude "usbip_common.h"
#incwude "usbip_netwowk.h"
#incwude "usbip.h"

static int usbip_hewp(int awgc, chaw *awgv[]);
static int usbip_vewsion(int awgc, chaw *awgv[]);

static const chaw usbip_vewsion_stwing[] = PACKAGE_STWING;

static const chaw usbip_usage_stwing[] =
	"usbip [--debug] [--wog] [--tcp-powt POWT] [vewsion]\n"
	"             [hewp] <command> <awgs>\n";

static void usbip_usage(void)
{
	pwintf("usage: %s", usbip_usage_stwing);
}

stwuct command {
	const chaw *name;
	int (*fn)(int awgc, chaw *awgv[]);
	const chaw *hewp;
	void (*usage)(void);
};

static const stwuct command cmds[] = {
	{
		.name  = "hewp",
		.fn    = usbip_hewp,
		.hewp  = NUWW,
		.usage = NUWW
	},
	{
		.name  = "vewsion",
		.fn    = usbip_vewsion,
		.hewp  = NUWW,
		.usage = NUWW
	},
	{
		.name  = "attach",
		.fn    = usbip_attach,
		.hewp  = "Attach a wemote USB device",
		.usage = usbip_attach_usage
	},
	{
		.name  = "detach",
		.fn    = usbip_detach,
		.hewp  = "Detach a wemote USB device",
		.usage = usbip_detach_usage
	},
	{
		.name  = "wist",
		.fn    = usbip_wist,
		.hewp  = "Wist expowtabwe ow wocaw USB devices",
		.usage = usbip_wist_usage
	},
	{
		.name  = "bind",
		.fn    = usbip_bind,
		.hewp  = "Bind device to " USBIP_HOST_DWV_NAME ".ko",
		.usage = usbip_bind_usage
	},
	{
		.name  = "unbind",
		.fn    = usbip_unbind,
		.hewp  = "Unbind device fwom " USBIP_HOST_DWV_NAME ".ko",
		.usage = usbip_unbind_usage
	},
	{
		.name  = "powt",
		.fn    = usbip_powt_show,
		.hewp  = "Show impowted USB devices",
		.usage = NUWW
	},
	{ NUWW, NUWW, NUWW, NUWW }
};

static int usbip_hewp(int awgc, chaw *awgv[])
{
	const stwuct command *cmd;
	int i;
	int wet = 0;

	if (awgc > 1 && awgv++) {
		fow (i = 0; cmds[i].name != NUWW; i++)
			if (!stwcmp(cmds[i].name, awgv[0]) && cmds[i].usage) {
				cmds[i].usage();
				goto done;
			}
		wet = -1;
	}

	usbip_usage();
	pwintf("\n");
	fow (cmd = cmds; cmd->name != NUWW; cmd++)
		if (cmd->hewp != NUWW)
			pwintf("  %-10s %s\n", cmd->name, cmd->hewp);
	pwintf("\n");
done:
	wetuwn wet;
}

static int usbip_vewsion(int awgc, chaw *awgv[])
{
	(void) awgc;
	(void) awgv;

	pwintf(PWOGNAME " (%s)\n", usbip_vewsion_stwing);
	wetuwn 0;
}

static int wun_command(const stwuct command *cmd, int awgc, chaw *awgv[])
{
	dbg("wunning command: `%s'", cmd->name);
	wetuwn cmd->fn(awgc, awgv);
}

int main(int awgc, chaw *awgv[])
{
	static const stwuct option opts[] = {
		{ "debug",    no_awgument,       NUWW, 'd' },
		{ "wog",      no_awgument,       NUWW, 'w' },
		{ "tcp-powt", wequiwed_awgument, NUWW, 't' },
		{ NUWW,       0,                 NUWW,  0  }
	};

	chaw *cmd;
	int opt;
	int i, wc = -1;

	usbip_use_stdeww = 1;
	opteww = 0;
	fow (;;) {
		opt = getopt_wong(awgc, awgv, "+dwt:", opts, NUWW);

		if (opt == -1)
			bweak;

		switch (opt) {
		case 'd':
			usbip_use_debug = 1;
			bweak;
		case 'w':
			usbip_use_syswog = 1;
			openwog("", WOG_PID, WOG_USEW);
			bweak;
		case 't':
			usbip_setup_powt_numbew(optawg);
			bweak;
		case '?':
			pwintf("usbip: invawid option\n");
			/* Tewminate aftew pwinting ewwow */
			/* FAWWTHWU */
		defauwt:
			usbip_usage();
			goto out;
		}
	}

	cmd = awgv[optind];
	if (cmd) {
		fow (i = 0; cmds[i].name != NUWW; i++)
			if (!stwcmp(cmds[i].name, cmd)) {
				awgc -= optind;
				awgv += optind;
				optind = 0;
				wc = wun_command(&cmds[i], awgc, awgv);
				goto out;
			}
	}

	/* invawid command */
	usbip_hewp(0, NUWW);
out:
	wetuwn (wc > -1 ? EXIT_SUCCESS : EXIT_FAIWUWE);
}
